File: PHP/show/Stage/AbstractClass.php
<?php

abstract class AbstractClass
{
    final public function templateMethod(): void
    {
        $this->baseOperation1();
        $this->requiredOperations1();
        $this->baseOperation2();
        $this->hook1();
        $this->requiredOperation2();
        $this->baseOperation3();
        $this->hook2();
    }

    protected function baseOperation1(): void
    {
        echo "AbstractClass says: I am doing the bulk of the work<br/>";
    }
    protected function baseOperation2(): void
    {
        echo "AbstractClass says: But I let subclasses override some operations<br/>";
    }

    protected function baseOperation3(): void
    {
        echo "AbstractClass says: But I am doing the bulk of the work anyway<br/>";
    }
    abstract protected function requiredOperations1(): void;

    abstract protected function requiredOperation2(): void;
}

File: Csharp/Behavioral/Template/Example2/Show/Show/AbstractClass.cs
using System;

namespace Show
{
    public abstract class AbstractClass
    {
        // The 'TemplateMethod' is the skeleton of an algorithm.
        // It is made final to prevent its structure from being altered.
        public void TemplateMethod()
        {
            BaseOperation();
            RequiredOperations1();
            RequiredOperation2();
            Hook1();
        }

        // A common operation used by the template method.
        void BaseOperation()
        {
            Console.WriteLine("BaseOperation: Common operation for all subclasses");
        }

        // These operations have to be implemented in subclasses.
        protected abstract void RequiredOperations1();
        protected abstract void RequiredOperation2();

        // Hooks provide additional extension points in some parts of the algorithm.
        protected virtual void Hook1() { }
    }
}